Fix: Another paragraph is inserted, inside InnerBlocks.

jorgefilipecosta opened the pull request:
Currently, we have a bug even if a block has a paragraph as the last block of the template if no locking exists another paragraph is inserted.

This happens because, before the template is processed the parent block is empty and DefaultBlockApppender is rendered right away. DefaultBlockApppender inserts blocks when it gets focused. The parent block has a mechanism to focus the first table and if InnerBlocks were not yet processed and the parent block does not contain inputs the DefaultBlockApppender is going to get the focus.

Before the template is processed we don't know what blocks will exist so during this phase DefaultBlockApppender should not be rendered. This PR makes sure that we don't render DefaultBlockApppender right after the block is inserted before the template is processed.

This behavior was noticed in https://github.com/WordPress/gutenberg/pull/9416.




## How has this been tested?
I added the test block available in https://gist.github.com/jorgefilipecosta/edafb2422ef41020d75619adf31d725e.

I checked that after inserting the first paragraph specified in the template gets focused and no other paragraph is created.

I checked the blocks depending on the behavior of DefaultBlockApppender inserting a block (columns) continue to work as before.


## Screenshots <!-- if applicable -->
test block: https://gist.github.com/jorgefilipecosta/edafb2422ef41020d75619adf31d725e
```
	var TEMPLATE = [
		[ 'core/paragraph', { placeholder: 'Paragraph 1', customFontSize: 35 } ],
	];
```
After:

<img width="568" alt="screen shot 2018-10-18 at 13 07 03" src="https://user-images.githubusercontent.com/11271197/47153548-e36a5b80-d2d7-11e8-83a6-33a9b39d97ce.png">

<img width="700" alt="screen shot 2018-10-18 at 13 06 54" src="https://user-images.githubusercontent.com/11271197/47153560-e9f8d300-d2d7-11e8-8086-cfa2e2252432.png">
Before:

<img width="675" alt="screen shot 2018-10-18 at 13 17 27" src="https://user-images.githubusercontent.com/11271197/47153679-30e6c880-d2d8-11e8-82b4-d4e7c19fa0b9.png">
<img width="656" alt="screen shot 2018-10-18 at 13 17 21" src="https://user-images.githubusercontent.com/11271197/47153683-33e1b900-d2d8-11e8-80da-7afc0294d311.png">

aduth reviewed (commented):
Do we need to wait  until `componentDidMount` to sync the template, or can it be done in `componentWillMount`?

Alternatively, can we not render `BlockList` until after the sync?

For what reason do we pass `template` to `BlockList`? The responsibilities are fuzzy to me, and makes me wonder if it's not `BlockList` to be synchronizing the template. But as I look to `BlockList`, it doesn't appear we _should_ be passing it (it's unused)?

Honestly, this feels like something where we're being punished for our non-ideal implementation of `InnerBlocks` template managing blocks state in response to lifecycle, rather than having more strict rendering control of blocks in a `BlocksList` (i.e. `value` or `blocks` property, similar to some ideas @youknowriad has expressed around `<Editor value={ ... } onChange={ ... } />`).

jorgefilipecosta committed:
Remove unused props passed to BlockList. (+2 squashed commits)
Squashed commits:
[23554fe7b] Fix: Another paragraph is inserted, inside the block even if the block template already includes a paragraph.
[a31f7b2ee] Update snapshot

tofumatt reviewed (dismissed):
I don't quite follow the code here, and I think having an E2E test that reproduced the testing instructions here would help to contextualise the code/issue better and also prevent regressions.

Could you add an E2E test for this?

jorgefilipecosta commented:
Hi @tofumatt thank you for reviewing this PR. We already had an end 2 end test case touching this, the snapshoot had the buggy implementation and this PR updated the snapshoot. But I agree we could benefit from a test dedicated to this behaviour. I added a new end 2 end test specifically for this behaviour.

tofumatt reviewed (approved):
Works for me! 👍
